File: solutions/cpp/pig-latin/2/pig_latin.cpp
#include "pig_latin.h"

#include <regex>
#include <iostream>

using namespace std;

namespace pig_latin {

    string translate(string english) {
        string pig("");
        regex words(R"(\w+)");
        sregex_iterator it(english.begin(), 
                     english.end(), words), end;
        for(;it!=end; it++) {            
            pig+=piggyfy((*it).str()) + " ";         
        }
        if(pig.back()==' ') 
            pig.pop_back();
        return pig;
    }

    
    string piggyfy(string e_word){
        regex vowels(R"(^([aeiou]|yt|xr)(\w*))");
        string cns_str[3] = {
            R"(([^aeiou]*[q][u])(\w*))",
            R"(([^aeiouy]+)(y\w*))", 
            R"(([^aeiou]+)([aeiou]\w*))"};
        smatch sm;
        if (regex_search(e_word, sm, vowels)) 
            return e_word+"ay";
        for (size_t i=0; i<3; i++){
            if(regex_match(e_word, sm, regex(cns_str[i])))
                return sm[2].str() + sm[1].str()+"ay";
        } 
        return e_word;
    }
    

}  // namespace pig_latin

File: solutions/dart/pig-latin/2/lib/pig_latin.dart
    String translate(String english){
     String pig = '';
     List<String> words = 
       RegExp(r"\w+").allMatches(english)
        .map((m) => m.group(0)).nonNulls.toList();
     for(var w in words){
        pig += piggifyWord(w) + ' ';
      }
      if(pig[pig.length-1]==' ')
        pig = pig.substring(0,pig.length-1);
      return pig;
    }

    String piggifyWord(String word){
      var vowels = RegExp(r"(^([aeiou]|yt|xr)(\w*))");
      var consonants = RegExp(r"^(?<move>sch|thr|[^aeiou]qu|ch|qu|rh|th|[^aeiuo])(?<keep>\w*)");
      if(vowels.hasMatch(word))
        return word + 'ay';
      if(consonants.hasMatch(word)){
          var match = consonants.firstMatch(word)!;
          return match.namedGroup('keep')! 
            + match.namedGroup('move')! + 'ay';
        }
      return word;
    }
  

